[FLINK-11501][Kafka Connector] Add ratelimiting to Kafka consumer

glaksh100 opened the pull request:
## What is the purpose of the change
This pull request adds a ratelimiting feature to the Flink Kafka consumer. There are instances when a Flink job that reads from Kafka can read at a significantly high throughput (particularly while processing a backlog) and degrade the underlying Kafka cluster. While Kafka quotas are perhaps the best way to enforce this ratelimiting, there are cases where such a setup is not available or easily enabled. In such a scenario, ratelimiting on the FlinkKafkaConsumer is a useful feature. 

## Brief change log
  - This feature is set by using a feature flag - `kafka.consumer.ratelimiting.enabled`
  -  A`RateLimiterFactory` is used to configure and create a Guava [RateLimiter](https://google.github.io/guava/releases/19.0/api/docs/index.html?com/google/common/util/concurrent/RateLimiter.html) with a desired rate.
  -  The `consumer.poll()` part of the `run()` loop in the `KafkaConsumerThread` is modularized into a `getRecordsFromKafka()` method.
  - The rate is controlled by setting the bytes received from Kafka as the parameter to the `acquire()` call. 
 
## Verifying this change
This change added tests and can be verified as follows:
 - Added a `testRateLimiting()` test in the `KafkaConsumerThreadTest` class.
 - Manually verified the change using a test application and screenshots of results are added [here](https://issues.apache.org/jira/browse/FLINK-11501?focusedCommentId=16762965&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16762965). 
- Added a `testRateLimitedConsumer()` to `Kafka09ITCase`

## Does this pull request potentially affect one of the following parts:
  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

becketqin commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaConsumerThread.java:
```diff
@@ -22,12 +22,16 @@
 import org.apache.flink.streaming.connectors.kafka.internals.KafkaTopicPartitionStateSentinel;
 import org.apache.flink.streaming.connectors.kafka.internals.metrics.KafkaMetricWrapper;
 
+import org.apache.flink.shaded.guava18.com.google.common.util.concurrent.RateLimiter;
```
Do we want to directly import class from the shaded jar?

tweise replied:
This import should not be here at all after we find the appropriate abstraction. But in general, we should use the shaded dependency or not?

zentol replied:
yes, if you want to use something from guava then you should use the shaded dependency. The only exception is if a another dependency exposes a guava-related API.

becketqin replied:
@zentol @tweise Just for my own curiosity. I agree that we should shade the jar to avoid interference with other user imports. What I am wondering is how should that be done. Whether it should be done in the pom.xml or from the import explicitly.  I am not sure about the difference between those two ways. But I did not see such explicit imports in Flink anywhere else. Am I missing something?

TisonKun replied:
@becketqin I'm not familiar with shaded dependency but let's say that it is a common case we use such explicit imports. You can just search "import org.apache.flink.shaded.guava18` and there are over 100 use points.

becketqin commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaConsumerThread.java:
```diff
@@ -482,6 +502,49 @@ void reassignPartitions(List<KafkaTopicPartitionState<TopicPartition>> newPartit
+	 * @param records List of ConsumerRecords.
+	 * @return Total batch size in bytes, including key and value.
+	 */
+	private int getRecordBatchSize(ConsumerRecords<byte[], byte[]> records) {
```
It is a little unfortunate that we have to make an additional iteration over the records to get the sizes. Another option is putting the throttling logic in the `AbstractFetcher`.

glaksh100 replied:
The reason for doing the iteration here was because we needed the record sizes prior to deserialization and I believe the records are already deserialized at the `AbstractFetcher` level?

becketqin replied:
The `AbstractFetcher` has two subclasses - `Kafka09Fetcher` and `Kafka08Fetcher`. The serialization is done inside them. 

`Kafka09Fetcher` is sort of an abstract fetcher for Kafka 0.9+. `Kafka08Fetcher` is more of a legacy. Maybe we can just do that in `Kafka09Fetcher`?

becketqin commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/config/RateLimiterFactory.java:
```diff
@@ -0,0 +1,101 @@
+/**
+ * A RateLimiterFactory that configures and creates a rate limiter.
+ */
+public class RateLimiterFactory {
```
Can we move this class to a separate package so it can be shared by all the other connectors?
For example, a `flink-connector-throttling` module in `flink-connectors`.

glaksh100 replied:
Yes, I think it makes sense to move this logic to a separate package within the `flink-connectors` module.  I will do that once we reach an agreement on the discussion. Thanks for suggesting.

glaksh100 replied:
I have not made this change yet. Perhaps we could do this as a part of augmenting the consumer and making the code more customizable (in another PR i.e.) ?

glaksh100 replied:
I opened a mailing list thread here to discuss where this module should live: https://lists.apache.org/thread.html/3de9d2353cf22aea0448fb744314103b5f88195216acc3bff449354a@%3Cdev.flink.apache.org%3E

glaksh100 replied:
This has now been moved to `org.apache.flink.api.common.io.ratelimiting`

becketqin commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/config/RateLimiterFactory.java:
```diff
@@ -0,0 +1,101 @@
+public class RateLimiterFactory {
+
+	/** Flag that indicates if ratelimiting is enabled. */
+	private static final String RATELIMITING_FLAG = "consumer.ratelimiting.enabled";
```
Personally, I prefer independent configurations if possible. I am wondering if the following configuration names would be clear enough:
```
CONSUMER_MAX_BYTES_PER_SECOND_CONFIG="consumer.max.bytes.per.second"
CONSUMER_MAX_RECORDS_PER_SECOND_CONFIG="consumer.max.records.per.second"
```
In the future we can add configurations for the producers.

The default value of the above configurations could be either be set to Long.MAX_VALUE, or -1 to indicate not throttling.

tweise replied:
There is no need for this configuration. When a rate limiter is set on the consumer, then it is "enabled", otherwise it is null and no rate limiting takes place.

glaksh100 replied:
Makes sense to enable ratelimiting based on whether or not a consumer's ratelimiter is null. To @becketqin's point, I think we would still need to pass the rate value itself from the consumer to its ratelimiter. Perhaps we could standardize that property across connectors?

glaksh100 replied:
I ended up passing the rate value to the setter that sets the ratelimiter on the consumer. Let me know if that looks alright.

tweise commented:
* @flinkbot approve description

tweise commented:
@flinkbot approve description

tweise commented on flink-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer010.java:
```diff
@@ -69,6 +70,10 @@
 	private static final long serialVersionUID = 2324564345203409112L;
 
+	/** Configuration to set consumer prefix for ratelimiting. **/
+	private static final String CONSUMER_PREFIX = "kafka";
```
There is not need for this prefix. Rate limiting should be configured per consumer. I have a more general comment on that.

glaksh100 replied:
Yeah, I think the consumer prefix is not adding any value here. Makes sense to not have it.

tweise commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/config/RateLimiterFactory.java:
```diff
@@ -0,0 +1,101 @@
+	 * @param runtimeContext Runtime context.
+	 * @param properties Consumer properties.
+	 */
+	public void configure(String consumerPrefix,
```
I would suggest to remove this property based configuration and have instead properties on the rate limiter that the user sets directly.

tweise commented:
@flinkbot approve consensus

tweise commented:
@rmetzger the flinkbot is awesome!

rmetzger commented:
Thanks a lot :) 
There's a cool new feature coming up in the pipeline: Labeling of the PR progress.

glaksh100 commented:
Thanks for taking a look @becketqin and @tweise. I think adding an interface that can extend to other connectors  and also not restricting consumers to use Guava's RateLimiter makes sense. Maybe something simple like below will work for the interface?

```
public interface FlinkConnectorRateLimiter {
	public void open(StreamingRuntimeContext runtimeContext);
	public <T> T create(long rate);
	public void setRate(long rate);
	public void close();
}
```
And connector specific RateLimiters can be defined to implement this interface. 

Also, point taken on keeping these discussions on the mailing list in the future.

glaksh100 commented:
@tweise @becketqin I updated the code to include a `FlinkConnectorRateLimiter` and also included a `DefaultKafkaRateLimiter` that implements the interface using Guava's RateLimiter. Let me know what you think.

tweise commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer09.java:
```diff
@@ -311,4 +323,19 @@ private static void setDeserializer(Properties props) {
+	 * @param kafkaRateLimiter
+	 * @param maxBytesPerSecond
+	 */
+	public void setRateLimiter(FlinkConnectorRateLimiter kafkaRateLimiter, long maxBytesPerSecond) {
```
Why the second parameter? That should be a setter on `DefaultKafkaRateLimiter`

glaksh100 replied:
Fixed.

tweise commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/config/DefaultKafkaRateLimiter.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * A default KafkaRateLimiter that uses Guava's RateLimiter to throttle the bytes read from Kafka.
+ */
+public class DefaultKafkaRateLimiter implements FlinkConnectorRateLimiter {
```
Since there is nothing Kafka specific here, but the implementation is specific to Guava, would `GuavaFlinkConnectorRateLimiter` be a better name?

glaksh100 replied:
I think that is a better name. Thanks for the suggestion.

tweise commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/config/FlinkConnectorRateLimiter.java:
```diff
@@ -0,0 +1,48 @@
+	void open(RuntimeContext runtimeContext);
+
+	/** Creates a rate limiter.  */
+	<T> T create();
```
This is confusing, the rate limiter was already created by the user.

Why is `acquire` not defined here?

glaksh100 replied:
This does look confusing. Essentially, we need `runtimeContext` to be initialized to be able to determine the per-subtask rate. Perhaps, only the `open()` method should suffice (and all related logic can be implemented there)? 

As for `acquire()` not being here, that was my bad. I have fixed this to include `acquire()` and completely removed the Guava import in the `KafkaConsumerThread` class (which was not the case before).

tweise replied:
It looks like `create` is redundant. The work done in that method can be moved to `open`.

glaksh100 replied:
Updated to only have `open()` and also moved the `create()` logic in the `GuavaFlinkConnectorRateLimiter` to reflect this.

tweise commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/config/FlinkConnectorRateLimiter.java:
```diff
@@ -0,0 +1,54 @@
+
+	long getRate();
+
+	void close();
```
The calls to close calls missing.

glaksh100 replied:
Thanks for catching. Updated to call `close()` in  the `KafkaConsumerThread` both in the `finally{}` block of the `run()` method and in the `shutdown()` method.

tweise commented on flink-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer010.java:
```diff
@@ -69,6 +70,12 @@
+	 * RateLimiter to throttle bytes read from Kafka. The rateLimiter is set via
+	 * {@link #setRateLimiter(FlinkConnectorRateLimiter)}.
+	 */
+	private FlinkConnectorRateLimiter rateLimiter;
```
Why repeat this here (already defined in `FlinkKafkaConsumer09`)?

glaksh100 replied:
Fair point. I removed the field and added a getter to access the `rateLimiter` from the super class. Let me know if that doesn't seem reasonable.

glaksh100 commented:
@tweise I added a test to the `Kafka09ITCase` that runs a job that uses the ratelimiter (and asserts an approximate rate). I also updated the `GuavaFlinkConnectorRateLimiter` to be serializable (which was a bug that I discovered on writing the IT case). Let me know your thoughts!

glaksh100 commented:
As per discussion on the [mailing list](https://lists.apache.org/thread.html/3de9d2353cf22aea0448fb744314103b5f88195216acc3bff449354a@%3Cdev.flink.apache.org%3E), I have moved `FlinkConnectorRateLimiter` and `GuavaFlinkConnectorRateLimiter` to `org.apache.flink.api.common.io.ratelimiting`. The latest commit includes changes to update references. @tweise PTAL when you have a chance and let me know what you think!

tweise commented on flink-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer010.java:
```diff
@@ -259,4 +267,5 @@ protected AbstractPartitionDiscoverer createPartitionDiscoverer(
 
 		return result;
 	}
+
```
remove

tweise commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer09.java:
```diff
@@ -311,4 +323,21 @@ private static void setDeserializer(Properties props) {
 	}
+
+
+	// ------------------------------------------------------------------------
```
Unless there is precedence let's remove this comment - it is pretty obvious what these methods are there for :)

glaksh100 replied:
Sounds good, removed :)

tweise commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaConsumerThread.java:
```diff
@@ -254,7 +264,8 @@ public void run() {
 					try {
-						records = consumer.poll(pollTimeout);
+						records = getRecordsFromKafka();
+
```
remove

tweise commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaConsumerThread.java:
```diff
@@ -316,6 +332,12 @@ public void shutdown() {
+		if (rateLimiter != null) {
+			rateLimiter.close();
+		}
+
```
remove

tweise commented on flink-connectors/flink-connector-kafka-0.9/src/test/java/org/apache/flink/streaming/connectors/kafka/Kafka09ITCase.java:
```diff
@@ -139,4 +159,118 @@ public void testCommitOffsetsToKafka() throws Exception {
+			public void invoke(String value, Context context) throws Exception {
+
+			}
+
```
remove

tweise commented on flink-connectors/flink-connector-kafka-0.9/src/test/java/org/apache/flink/streaming/connectors/kafka/Kafka09ITCase.java:
```diff
@@ -139,4 +159,118 @@ public void testCommitOffsetsToKafka() throws Exception {
+		Assert.assertNotNull(consumer09.getRateLimiter());
+		Assert.assertEquals(consumer09.getRateLimiter().getRate(), globalRate);
+
+    	// Approximate bytes/second read based on job execution time.
```
indent

tweise commented on flink-connectors/flink-connector-kafka-0.9/src/test/java/org/apache/flink/streaming/connectors/kafka/Kafka09ITCase.java:
```diff
@@ -139,4 +159,118 @@ public void testCommitOffsetsToKafka() throws Exception {
+    	// Approximate bytes/second read based on job execution time.
+		long bytesPerSecond = 100 * 1000L / (endTime - startTime);
+		Assert.assertTrue(bytesPerSecond > 0);
+		Assert.assertTrue(bytesPerSecond <= globalRate);
```
Thanks for writing this additional test. Just to confirm, would this test generate a higher rate without the limit?

glaksh100 replied:
Yes. That is correct. The rate is close to 45 bytes/second when run without setting the ratelimiter.

tweise replied:
OK, that should be good. I would make a note of that in the test though. What makes this a bit shaky is the fact that you could get a false positive when the initialization/execution overhead is just enough to hide a faster read. I'm not sure how that can be done in a much better way though. Perhaps measure the time in the deserializer? Otherwise we can leave it as is.

glaksh100 replied:
I added details to the javadoc with the test expectation. Moving it to deserializer is a good idea. I updated the test. The gain is ~ 200 ms (not all that much but closer than what I had before). Thanks for the suggestion.

tweise commented on flink-connectors/flink-connector-kafka-0.9/src/test/java/org/apache/flink/streaming/connectors/kafka/Kafka09ITCase.java:
```diff
@@ -139,4 +159,118 @@ public void testCommitOffsetsToKafka() throws Exception {
+		Assert.assertTrue(bytesPerSecond > 0);
+		Assert.assertTrue(bytesPerSecond <= globalRate);
+		deleteTestTopic(testTopic);
+
```
remove

tweise commented on flink-core/src/main/java/org/apache/flink/api/common/io/ratelimiting/FlinkConnectorRateLimiter.java:
```diff
@@ -0,0 +1,51 @@
+@PublicEvolving
+public interface FlinkConnectorRateLimiter {
+
+	void open(RuntimeContext runtimeContext);
```
javadoc - mention what this method would be useful for

glaksh100 replied:
Added a javadoc.

tweise commented on flink-core/src/main/java/org/apache/flink/api/common/io/ratelimiting/GuavaFlinkConnectorRateLimiter.java:
```diff
@@ -0,0 +1,81 @@
+/**
+ * An implemetation of {@link FlinkConnectorRateLimiter} that uses Guava's RateLimiter for rate limiting.
+ */
+public class GuavaFlinkConnectorRateLimiter implements FlinkConnectorRateLimiter, Serializable {
```
FlinkConnectorRateLimiter must be serializable, so we should move that up.

glaksh100 replied:
Moved up.

tweise commented on flink-connectors/flink-connector-kafka-0.9/src/test/java/org/apache/flink/streaming/connectors/kafka/Kafka09ITCase.java:
```diff
@@ -139,4 +159,125 @@ public void testCommitOffsetsToKafka() throws Exception {
+
+			}
+		});
+		long startTime = System.currentTimeMillis();
```
remove this and `endTime` below

tweise commented:
@flinkbot approve all

tweise commented:
@glaksh100 thanks for the contribution!

glaksh100 commented:
@tweise Thank you for the patient and thorough review :)

FaxianZhao commented on flink-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka010Fetcher.java:
```diff
@@ -75,7 +77,7 @@ public Kafka010Fetcher(
 				subtaskMetricGroup,
 				consumerMetricGroup,
-				useMetrics);
+				useMetrics, rateLimiter);
```
It seems you break the code style. `rateLimiter);` should in the next line.
